README
NAME
    Data::Timeline - Time line represented as an object

VERSION
    version 1.100850

SYNOPSIS
        my $timeline = Data::Timeline->new;
        $timeline->entries_push(Data::Timeline::Entry->new(
            # ...
        ));
        $timeline->merge_timeline($other_timeline);

DESCRIPTION
    This class represents a time line, which is a collection of time line
    entry objects (see Data::Timeline::Entry).

METHODS
  filter_timeline_by_date
    FIXME

  filter_timeline_by_type
    FIXME

  merge_timeline
    FIXME

INSTALLATION
    See perlmodinstall for information and options on installing Perl
    modules.

BUGS AND LIMITATIONS
    No bugs have been reported.

    Please report any bugs or feature requests through the web interface at
    <http://rt.cpan.org/Public/Dist/Display.html?Name=Data-Timeline>.

AVAILABILITY
    The latest version of this module is available from the Comprehensive
    Perl Archive Network (CPAN). Visit <http://www.perl.com/CPAN/> to find a
    CPAN site near you, or see <http://search.cpan.org/dist/Data-Timeline/>.

    The development version lives at
    <http://github.com/hanekomu/Data-Timeline/>. Instead of sending patches,
    please fork this project using the standard git and github
    infrastructure.

AUTHOR
      Marcel Gruenauer <marcel@cpan.org>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2007 by Marcel Gruenauer.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Revision history for Perl extension Data-Timeline

1.100850  2010-03-26 16:42:38 Europe/Vienna
    - converted the distribution to Dist::Zilla-style

0.04  Thu Jul 25 15:27:43 CEST 2008 (Marcel Gruenauer <marcel@cpan.org>)
    - Changed dist style and Changes back to standard. People didn't like it -
      the nail that sticks out gets hammered down.
    - Added standard test files; this will also help with CPANTS.

0.03  Thu, 29 May 2008 12:03:57 -0000 (Marcel Gruenauer <marcel@cpan.org>)
    - set the version to 0.03
    - moved generated documentation of interited methods to after documentation
      of this class' methods so the manpage is more readable
    - .shipit: fixed svk.tagpattern

0.02  Wed, 21 Nov 2007 18:22:19 -0000 (Marcel Gruenauer <marcel@cpan.org>)
    - Removed superfluous requirement for DateTime::Format::DateParse
    - Makefile.PL: moved auto_install call to after use_standard_tests call
    - .shipit: added Twitter step
    - Makefile.PL: removed comment
    - updated MANIFEST
    - updated t/perlcriticrc
    - tags: BUGFIX

0.01  Mon, 12 Nov 2007 22:51:10 -0000 (Marcel Gruenauer <marcel@cpan.org>)
    - original version


lib/Data/Timeline.pm
use 5.008;
use strict;
use warnings;

package Data::Timeline;
our $VERSION = '1.100850';
# ABSTRACT: Time line represented as an object
use 5.006;
use parent qw(Class::Accessor::Complex Class::Accessor::Constructor);
__PACKAGE__
    ->mk_constructor
    ->mk_array_accessors(qw(entries));

sub merge_timeline {
    my ($self, $timeline) = @_;
    $self->entries(sort { $a->timestamp <=> $b->timestamp }
          ($self->entries, $timeline->entries));
    $self;    # for chaining
}

sub filter_timeline_by_type {
    my ($self, $type) = @_;
    Data::Timeline->new(entries => grep { $_->type eq $type } $self->entries);
}

sub filter_timeline_by_date {
    my ($self, $from, $to) = @_;
    Data::Timeline->new(
        entries => grep { ($from <= $_->timestamp) && ($_->timestamp <= $to) }
          $self->entries);
}
1;


__END__
=pod

=for test_synopsis my $other_timeline;

=head1 NAME

Data::Timeline - Time line represented as an object

=head1 VERSION

version 1.100850

=head1 SYNOPSIS

    my $timeline = Data::Timeline->new;
    $timeline->entries_push(Data::Timeline::Entry->new(
        # ...
    ));
    $timeline->merge_timeline($other_timeline);

=head1 DESCRIPTION

This class represents a time line, which is a collection of time line entry
objects (see L<Data::Timeline::Entry>).

=head1 METHODS

=head2 filter_timeline_by_date

FIXME

=head2 filter_timeline_by_type

FIXME

=head2 merge_timeline

FIXME

=head1 INSTALLATION

See perlmodinstall for information and options on installing Perl modules.

=head1 BUGS AND LIMITATIONS

No bugs have been reported.

Please report any bugs or feature requests through the web interface at
L<http://rt.cpan.org/Public/Dist/Display.html?Name=Data-Timeline>.

=head1 AVAILABILITY

The latest version of this module is available from the Comprehensive Perl
Archive Network (CPAN). Visit L<http://www.perl.com/CPAN/> to find a CPAN
site near you, or see
L<http://search.cpan.org/dist/Data-Timeline/>.

The development version lives at
L<http://github.com/hanekomu/Data-Timeline/>.
Instead of sending patches, please fork this project using the standard git
and github infrastructure.

=head1 AUTHOR

  Marcel Gruenauer <marcel@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2007 by Marcel Gruenauer.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut



lib/Data/Timeline/Entry.pm
use 5.008;
use strict;
use warnings;

package Data::Timeline::Entry;
our $VERSION = '1.100850';
# ABSTRACT: A time line entry
use parent qw(Class::Accessor::Complex Class::Accessor::Constructor);
__PACKAGE__
    ->mk_constructor
    ->mk_scalar_accessors(qw(timestamp type description));
1;


__END__
=pod

=head1 NAME

Data::Timeline::Entry - A time line entry

=head1 VERSION

version 1.100850

=head1 SYNOPSIS

    my $entry = Data::Timeline::Entry->new(
        timestamp   => '...',
        type        => '...',
        description => '...',
    );

=head1 DESCRIPTION

This class represents a time line entry. An entry has a timestamp, a type and
a description. The timestamp needs to be a L<DateTime> object. The type is
freely definable, but you need to be consistent. If you want to display
multiple time lines side-by-side you probably need to define the entry type
for each column. The description is a string that says what this entry is
about.

=head1 INSTALLATION

See perlmodinstall for information and options on installing Perl modules.

=head1 BUGS AND LIMITATIONS

No bugs have been reported.

Please report any bugs or feature requests through the web interface at
L<http://rt.cpan.org/Public/Dist/Display.html?Name=Data-Timeline>.

=head1 AVAILABILITY

The latest version of this module is available from the Comprehensive Perl
Archive Network (CPAN). Visit L<http://www.perl.com/CPAN/> to find a CPAN
site near you, or see
L<http://search.cpan.org/dist/Data-Timeline/>.

The development version lives at
L<http://github.com/hanekomu/Data-Timeline/>.
Instead of sending patches, please fork this project using the standard git
and github infrastructure.

=head1 AUTHOR

  Marcel Gruenauer <marcel@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2007 by Marcel Gruenauer.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut



lib/Data/Timeline/Builder.pm
use 5.008;
use strict;
use warnings;

package Data::Timeline::Builder;
our $VERSION = '1.100850';
# ABSTRACT: Base class for time line builders
use Data::Timeline;
use Data::Timeline::Entry;
use parent qw(Class::Accessor::Complex Class::Accessor::Constructor);
__PACKAGE__
    ->mk_constructor
    ->mk_abstract_accessors(qw(create));

sub make_timeline {
    my $self = shift;
    Data::Timeline->new(@_);
}

sub make_entry {
    my $self = shift;
    Data::Timeline::Entry->new(@_);
}
1;


__END__
=pod

=for test_synopsis my @entries;

=head1 NAME

Data::Timeline::Builder - Base class for time line builders

=head1 VERSION

version 1.100850

=head1 SYNOPSIS

    Data::Timeline::Builder->new;

    package Data::Timeline::SVK;
    use base 'Data::Timeline::Builder';

    sub create {
        my $self = shift;
        my $timeline = $self->make_timeline;
        for (@entries) {
            # ...
            $timeline->entries_push($self->make_entry(
                timestamp   => '...',
                description => '...',
                type        => 'svk',
            ));
        }
        $timeline;
    }

=head1 DESCRIPTION

This is a base class for time line builders. Subclasses need to implement the
C<create()> method.

=head1 METHODS

=head2 make_entry

FIXME

=head2 make_timeline

FIXME

=head1 INSTALLATION

See perlmodinstall for information and options on installing Perl modules.

=head1 BUGS AND LIMITATIONS

No bugs have been reported.

Please report any bugs or feature requests through the web interface at
L<http://rt.cpan.org/Public/Dist/Display.html?Name=Data-Timeline>.

=head1 AVAILABILITY

The latest version of this module is available from the Comprehensive Perl
Archive Network (CPAN). Visit L<http://www.perl.com/CPAN/> to find a CPAN
site near you, or see
L<http://search.cpan.org/dist/Data-Timeline/>.

The development version lives at
L<http://github.com/hanekomu/Data-Timeline/>.
Instead of sending patches, please fork this project using the standard git
and github infrastructure.

=head1 AUTHOR

  Marcel Gruenauer <marcel@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2007 by Marcel Gruenauer.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut



lib/Data/Timeline/Formatter.pm
use 5.008;
use strict;
use warnings;

package Data::Timeline::Formatter;
our $VERSION = '1.100850';
# ABSTRACT: Base class for time line formatters
use parent qw(Class::Accessor::Complex Class::Accessor::Constructor);
__PACKAGE__
    ->mk_constructor
    ->mk_abstract_accessors(qw(format));
1;


__END__
=pod

=head1 NAME

Data::Timeline::Formatter - Base class for time line formatters

=head1 VERSION

version 1.100850

=head1 SYNOPSIS

    package Data::Timeline::My::Formatter;
    use base 'Data::Timeline::Formatter';

    sub format {
        # ...
    }

=head1 DESCRIPTION

This class is a base class for formatters. Subclasses need to implement the
format() method, which takes a time line object and outputs it in some way.

=head1 INSTALLATION

See perlmodinstall for information and options on installing Perl modules.

=head1 BUGS AND LIMITATIONS

No bugs have been reported.

Please report any bugs or feature requests through the web interface at
L<http://rt.cpan.org/Public/Dist/Display.html?Name=Data-Timeline>.

=head1 AVAILABILITY

The latest version of this module is available from the Comprehensive Perl
Archive Network (CPAN). Visit L<http://www.perl.com/CPAN/> to find a CPAN
site near you, or see
L<http://search.cpan.org/dist/Data-Timeline/>.

The development version lives at
L<http://github.com/hanekomu/Data-Timeline/>.
Instead of sending patches, please fork this project using the standard git
and github infrastructure.

=head1 AUTHOR

  Marcel Gruenauer <marcel@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2007 by Marcel Gruenauer.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut



lib/Data/Timeline/Formatter/HTML.pm
use 5.008;
use strict;
use warnings;

package Data::Timeline::Formatter::HTML;
our $VERSION = '1.100850';
# ABSTRACT: Print time line entry types side-by-side in an HTML table
use HTML::Table;
use parent qw(Data::Timeline::Formatter);
__PACKAGE__->mk_array_accessors(qw(columns));

sub format {
    my ($self, $timeline) = @_;
    my $table = HTML::Table->new(-head => [ 'timestamp', $self->columns ],);
    for my $entry ($timeline->entries) {
        my @row = (sprintf "%s" => $entry->timestamp);
        for my $col_type ($self->columns) {
            if ($entry->type eq $col_type) {
                push @row => $entry->description;
            } else {
                push @row => '';
            }
        }
        $table->addRow(@row);
    }
    $table->print;
}
1;


__END__
=pod

=for test_synopsis my $timeline;

=head1 NAME

Data::Timeline::Formatter::HTML - Print time line entry types side-by-side in an HTML table

=head1 VERSION

version 1.100850

=head1 SYNOPSIS

    Data::Timeline::Formatter::HTML->new(
        columns => [ qw(iscrobbler svk) ],
    )->format($timeline);

=head1 DESCRIPTION

This class is a time line formatter. It takes a time line containing entries
of one or more entry types and a column definition. The column definition says
for each column which type of entries it should contain. The formatter's
C<format()> method will then print a simple HTML table containing the
requested columns, with a column for the timestamp at the beginning.

The column definition is a list of entry type strings. pairs. So for the
example in the synopsis, the first column would contain the timestamp, the
second column would contain C<iscrobbler> entries, produced by
L<Data::Timeline::IScrobbler>, and the third column would contain C<svk>
entries, produced by L<Data::Timeline::SVK>.

=head1 METHODS

=head2 format

FIXME

=head1 INSTALLATION

See perlmodinstall for information and options on installing Perl modules.

=head1 BUGS AND LIMITATIONS

No bugs have been reported.

Please report any bugs or feature requests through the web interface at
L<http://rt.cpan.org/Public/Dist/Display.html?Name=Data-Timeline>.

=head1 AVAILABILITY

The latest version of this module is available from the Comprehensive Perl
Archive Network (CPAN). Visit L<http://www.perl.com/CPAN/> to find a CPAN
site near you, or see
L<http://search.cpan.org/dist/Data-Timeline/>.

The development version lives at
L<http://github.com/hanekomu/Data-Timeline/>.
Instead of sending patches, please fork this project using the standard git
and github infrastructure.

=head1 AUTHOR

  Marcel Gruenauer <marcel@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2007 by Marcel Gruenauer.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut



lib/Data/Timeline/Formatter/SimpleTable.pm
use 5.008;
use strict;
use warnings;

package Data::Timeline::Formatter::SimpleTable;
our $VERSION = '1.100850';
# ABSTRACT: Print time line entry types side-by-side in a text table
use Text::SimpleTable;
use parent qw(Data::Timeline::Formatter);
__PACKAGE__->mk_array_accessors(qw(columns));

sub format {
    my ($self, $timeline) = @_;
    my $table = Text::SimpleTable->new([ 20, 'timestamp' ], $self->columns,);
    my @col_types = map { $_->[1] } $self->columns;
    for my $entry ($timeline->entries) {
        my @row = (sprintf "%s" => $entry->timestamp);
        for my $col_type (@col_types) {
            if ($entry->type eq $col_type) {
                push @row => $entry->description;
            } else {
                push @row => '';
            }
        }
        $table->row(@row);
    }
    print $table->draw;
}
1;


__END__
=pod

=for test_synopsis my $timeline;

=head1 NAME

Data::Timeline::Formatter::SimpleTable - Print time line entry types side-by-side in a text table

=head1 VERSION

version 1.100850

=head1 SYNOPSIS

    Data::Timeline::Formatter::SimpleTable->new(
        columns => [ [ 40, 'iscrobbler' ], [ 40, 'svk' ] ],
    )->format($timeline);

=head1 DESCRIPTION

This class is a time line formatter. It takes a time line containing entries
of one or more entry types and a column definition. The column definition says
for each column how wide it should be and which type of entries it should
contain. The formatter's C<format()> method will then print a simple text
table containing the requested columns, with a column for the timestamp at the
beginning.

The column definition is a list of C<[ $width, $type ]> pairs. So for the
example in the synopsis, the first column would contain the timestamp, the
second column would be 40 characters wide and contain C<iscrobbler> entries,
produced by L<Data::Timeline::IScrobbler>, and the third column would also
be 40 characters wide and contain C<svk> entries, produced by
L<Data::Timeline::SVK>.

=head1 METHODS

=head2 format

FIXME

=head1 INSTALLATION

See perlmodinstall for information and options on installing Perl modules.

=head1 BUGS AND LIMITATIONS

No bugs have been reported.

Please report any bugs or feature requests through the web interface at
L<http://rt.cpan.org/Public/Dist/Display.html?Name=Data-Timeline>.

=head1 AVAILABILITY

The latest version of this module is available from the Comprehensive Perl
Archive Network (CPAN). Visit L<http://www.perl.com/CPAN/> to find a CPAN
site near you, or see
L<http://search.cpan.org/dist/Data-Timeline/>.

The development version lives at
L<http://github.com/hanekomu/Data-Timeline/>.
Instead of sending patches, please fork this project using the standard git
and github infrastructure.

=head1 AUTHOR

  Marcel Gruenauer <marcel@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2007 by Marcel Gruenauer.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut



Makefile.PL

use strict;
use warnings;

 BEGIN { require 5.008; } 

use ExtUtils::MakeMaker 6.11;



my %WriteMakefileArgs = (
                       'test' => {
                                   'TESTS' => 't/*.t'
                                 },
                       'NAME' => 'Data::Timeline',
                       'DISTNAME' => 'Data-Timeline',
                       'CONFIGURE_REQUIRES' => {
                                                 'ExtUtils::MakeMaker' => '6.11'
                                               },
                       'AUTHOR' => 'Marcel Gruenauer <marcel@cpan.org>',
                       'BUILD_REQUIRES' => {},
                       'ABSTRACT' => 'Time line represented as an object',
                       'EXE_FILES' => [],
                       'VERSION' => '1.100850',
                       'PREREQ_PM' => {
                                        'Class::Accessor::Complex' => '0',
                                        'Scalar::Util' => '0',
                                        'parent' => '0',
                                        'English' => '0',
                                        'HTML::Table' => '0',
                                        'Text::SimpleTable' => '0',
                                        'Carp' => '0',
                                        'Test::More' => '0.88',
                                        'Class::Accessor::Constructor' => '0',
                                        'File::Find' => '0',
                                        'File::Temp' => '0'
                                      },
                       'LICENSE' => 'perl'
                     );


delete $WriteMakefileArgs{LICENSE}
  unless eval { ExtUtils::MakeMaker->VERSION(6.31) };

WriteMakefile(%WriteMakefileArgs);



